### PR TITLE
SCI32: Fix dynamic line creation

### PR DESCRIPTION
**Pull request by bluegr**
This makes the crystal puzzle in Torin (room 43000) work correctly.
The results have been verified with LSL6 as well

- Fix the screen item rect to be the one returned by makeLineBitmap.
  This fixes line placement
- Improve sanity check in the line plotter to work with coordinates
  instead of bitmap buffer size
- Fix off-by-one error in the line screenitem rectangle
- Don't create a draw rectangle from the line coordinates, as these
  may not form a valid rectangle

### CODE REVIEW

**Comment by csnover** [comment]
The game crashes with this patch applied with this crystal configuration:

screen shot 2016-08-24 at 09 57 03

The stack trace shows an attempted write to y=-1.

File: engines/sci/graphics/paint32.cpp [hunk 1]
```diff
@@ -113,24 +112,26 @@ void GfxPaint32::plotter(int x, int y, int color, void *data) {
...
-	outRect.bottom = (startPoint.y > endPoint.y ? startPoint.y : endPoint.y) + halfThickness + 1;
+	outRect.left   = MIN<int16>(startPoint.x, endPoint.x) - halfThickness;
+	outRect.top    = MIN<int16>(startPoint.y, endPoint.y) - halfThickness;
+	outRect.right  = MAX<int16>(startPoint.x, endPoint.x) + (halfThickness + 1) + 1;	// round up halfThickness
+	outRect.bottom = MAX<int16>(startPoint.y, endPoint.y) + (halfThickness + 1) + 1;	// round up halfThickness
 
 	reg_t bitmapId;
-	SciBitmap &bitmap = *_segMan->allocateBitmap(&bitmapId, outRect.width(), outRect.height(), skipColor, 0, 0, g_sci->_gfxFrameout->getCurrentBuffer().scriptWidth, g_sci->_gfxFrameout->getCurrentBuffer().scriptHeight, 0, false, true);
+	uint16 scaledWidth = g_sci->_gfxFrameout->getCurrentBuffer().scriptWidth;
+	uint16 scaledHeight = g_sci->_gfxFrameout->getCurrentBuffer().scriptHeight;
```
Comment (csnover) [inline engines/sci/graphics/paint32.cpp]: Mismatched variable names.

Comment (bluegr) [inline engines/sci/graphics/paint32.cpp]: Are you referring to spacing?

**Comment by bluegr** [comment]
Interesting, thanks for finding that! Will test later on today

File: engines/sci/graphics/paint32.cpp (hunk 1 above)
Comment (csnover) [inline engines/sci/graphics/paint32.cpp]: No, I am referring to the mismatched variable names “scaled” and “script”.

File: engines/sci/graphics/paint32.cpp [hunk 2]
```diff
@@ -152,12 +153,11 @@ reg_t GfxPaint32::makeLineBitmap(const Common::Point &startPoint, const Common::
...
-		startPoint.x - outRect.left,
-		startPoint.y - outRect.top,
-		endPoint.x - outRect.left,
-		endPoint.y - outRect.top
-	);
+	// Change coordinates to be relative to the bitmap
+	int16 x1 = startPoint.x - outRect.left;
+	int16 y1 = startPoint.y - outRect.top;
+	int16 x2 = endPoint.x - outRect.left;
+	int16 y2 = endPoint.y - outRect.top;
```
Comment (csnover) [inline engines/sci/graphics/paint32.cpp]: Const correctness throughout would be appreciated.

File: engines/sci/graphics/paint32.cpp [hunk 3]
```diff
@@ -113,24 +112,26 @@ void GfxPaint32::plotter(int x, int y, int color, void *data) {
 			properties.patternIndex = 0;
 		}
 	} else {
-		warning("GfxPaint32::plotter: Attempted to write out of bounds (%u >= %u)", index, properties.bitmap->getDataSize());
+		warning("GfxPaint32::plotter: Attempted to write out of bounds (%u, %u) - bounds are (%u, %u)", x, y, properties.bitmap->getWidth(), properties.bitmap->getHeight());
```
Comment (csnover) [inline engines/sci/graphics/paint32.cpp]: This warns a lot; I wonder if there is another defect somewhere else causing overdraw.

**Comment by bluegr** [comment]
All the issues mentioned have been addressed now. I've checked the behavior against the original, and the results are the same. Check a reference video here:

https://www.youtube.com/watch?v=BIDT0A_6cdQ#t=1h46m17s

**Comment by csnover** [comment]
The wide line is not drawing at the correct position when slope = 0 or ∞.

untitled-2

**Comment by bluegr** [comment]
Thanks for the feedback, the issue was with the way our thick line implementation differs from Sierra's. The issue in the picture has been fixed now

**Comment by bluegr** [comment]
Update: Verified with @sev that the difference in our thick line implementation was unintentional, so the fixes for thick line drawing in this pull request have been pushed to common code

**Comment by csnover** [comment]
There are still some problems with the line drawing, most noticeably where slope = 1 (lower-right).

untitled-3

**Comment by bluegr** [comment]
Looks like a similar bug to the one above, but this one will be harder to fix. Will check it out

**Comment by bluegr** [comment]
The problem above seems to be an issue with thick line drawing, where an extra padding of 1 pixel is added in this case.

We are using the same thick line algorithm as SSCI, Bresenham by Foley & Van Dam, and have taken our thick line drawing algorithm from gd (function gdImageLine):
https://github.com/libgd/libgd/blob/master/src/gd.c

Line thickness in this case is 3 for thick lines, and 1 for normal lines.

This will need more investigation...

**Comment by csnover** [comment]
Addressed by PR #835.

**Comment by bluegr** [comment]
The commit that has been added in PR #835 is an earlier version of the final commit in this PR. I'll update master with the changes that I made.

**Pull request closed without merging**